### PR TITLE
MQ Performance Test Template

### PR DESCRIPTION
**Pull request by jmbeach**
This is a template inspired by the base test that is geared towards analyzing performance of MQ’s.

I wrote this on my phone so forgive typos and such

### CODE REVIEW

File: DesignOfExperiments/Performance.md [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+1,1617516000,1617516005
+2,1617516010,1617516015
+...
+```
+
+Link to your program’s usage here:
+
+[Replace me](Replace me)
+
+### Data Processor
```
Comment (TheENGR) [inline DesignOfExperiments/Performance.md:78]: If everyone generates proper data, can we have a single combiner that the DoE team created to parse it? Saves reinventing the wheel.

Comment (jmbeach) [reply DesignOfExperiments/Performance.md:78]: For sure

**Review by laynemoseley — DISMISSED** [review]

**Comment by laynemoseley** [comment]
if you really did write this on your phone, _kudos_. That is a lot of markdown for a phone.

**Comment by laynemoseley** [comment]
@jmbeach my main feedback is that as the tests scale up, we should limit the number of variables that are changes. For example, increasing message count _and_ message size at the same time won't give a clear picture of where the bottlenecks are.

For example, we could have 2 types of bottleneck tests:

- throughput
- message size

where each of those have their own test and they scale up individually.

Am I making sense?

**Comment by laynemoseley** [comment]
The other test I've been curious about as well with MQ if fail scenarios. For example, if you have redundancy producers, how many messages are lost when the lead producer fails and a new one takes its place?

**Comment by jmbeach** [comment]
@laynemoseley about limiting the changing variables. I agree, but I also worry about assigning too much work to the test runners. I actually would lean towards completely fixing some of the variables and never changing them

**Comment by laynemoseley** [comment]
which ones would you make static?

**Comment by laynemoseley** [comment]
we could have two test runs, one that _just_ increases message volume, and one that _just_ increases message size, for example:

Message Volume:

- 1000 messages per second
- 10000 messages per second
- 100000 messages per second

Message Size (static 1000 messages per second)
- 10 bytes
- 100 bytes
- 1000 bytes

etc, etc, etc

**Comment by thesammiller** [comment]
Thanks for this. I like a lot of what you guys are saying. 

In the interest of controlling variables, should we also split up single vs. multi threading? I'm curious if there may be a lot of variation in the implementations multi-threading and if it'd be helpful to limit to a single thread for a baseline.

**Comment by jmbeach** [comment]
My intention with multithreading is to prevent the user from having to spin up 100 VM's. Instead, they can maximize the utilization of cores on 5 VM's.

**Comment by jmbeach** [comment]
@laynemoseley that might be a better idea to adjust volume like you're saying. then we could just have the publisher run for a fixed duration and the consumer run until messages are all consumed

**Comment by jmbeach** [comment]
As I mentioned in [my discussion post](https://github.com/CS-6381/FinalProject/discussions/4), I'm basing this heavily off of [this article](https://softwaremill.com/mqperf/). They adjusted:
* instances of producer programs
* instances of consumer programs
* number of threads in producer program
* number of threads in consumer program

That's actually it. They must have ran their programs for fixed durations as well. So I'd be good with the following:

* Make message size static
* Remove specified message counts
    * Just run for a specific duration and see how many messages end up getting sent (what is the end throughput)
* Remove batching
    * It would be good to see it, but it's going to over-complicate things

Then each test just adjusts:
* Producer nodes
* Consumer nodes
* Producer threads
* Consumer threads

Just like the artice

File: DesignOfExperiments/Performance.md [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+The program should send 1,000 messages and then terminate once all messages are sent and all acknowledgements are received.
+
+Link to your program’s usage here:
+
+[Replace me](Replace me)
+
+### Consumer Program
+
+Design a **python** program utilizing your assigned MQ technology that consumes as many messages per second as possible (For example: `while true: consume_essage()`)
```
Comment (TheENGR) [inline DesignOfExperiments/Performance.md:42]:
```suggestion
Design a **python** program utilizing your assigned MQ technology that consumes as many messages per second as possible (For example: `while true: consume_message()`)
```

**Review by laynemoseley — APPROVED** [review]

**Pull request merged**